### PR TITLE
Convert to IIncrementalGenerator

### PR DESCRIPTION
**Pull request by davhdavh**
Tried to run it on a project with a couple of thousand resx files and ran into a couple of issues.
1. Dont generate invalid namespace for root level resx files
2. Dont generate invalid classnames when resx files have multiple extensions
3. Give compiler warning on duplicate members
4. Give compiler warning when member has same name as class
5. Convert to IIncrementalGenerator + instant intellisense when changing file

### CODE REVIEW

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs]: Please apply the `sealed` modifier.

File: VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs]: Can you please explain why you replaced `MemoryStream` with `StringReader`? I guess because of performance reasons?

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs]: Can you please put these properties before the constructor?

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/GlobalOptions.cs]: Can you please put these properties before the constructor?

File: VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj]: If there are no breaking changes in this PR, I think bumping the version number to either `2.0.3` or `2.1.0` makes more sense.

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:31]: Why is the `Valid` property needed? I can see this field is set only once here.

File: VocaDb.ResXFileCodeGenerator/SourceGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/SourceGenerator.cs]: Style nit: Please place spaces before and after the `=>` operator.

File: VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs]: Style nit: Please place spaces before and after the `=>` operator.

File: VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs]: Style nit: Please place spaces before and after the `=>` operator.

File: VocaDb.ResXFileCodeGenerator/SourceGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/SourceGenerator.cs]: Please replace `== null` with `is null`.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: Can you please prefix private static fields with `s_`?

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/GlobalOptions.cs]: Please apply the `sealed` modifier.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: old java dev? It doesn't matter in C# since the default is non-virtual. I have applied it nonetheless.

File: VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs]: Because the built-in method of reading the file uses SourceText object, and there was no particular reason to convert to utf8 bytes just to convert it back a couple of lines later.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: done

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/GlobalOptions.cs]: done

File: VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs]: done

File: VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs]: done

File: VocaDb.ResXFileCodeGenerator/SourceGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/SourceGenerator.cs]: done

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: done

File: VocaDb.ResXFileCodeGenerator/SourceGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/SourceGenerator.cs]: done

File: VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj]: True, there are however substantial speed increases and the generation is based on what the editor says the file contains rather than the file content. I personally still think it should be a 3.0.
Also, I am adding the partial class generation we discussed in issue #5, that will be included in next commit.

**Comment by davhdavh** [comment]
I didn't actually test this yet, so just review it, don't merge it.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: I would use named arguments for better readability here, but I can do it after merging this PR.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: I think the null-forgiving operator is not needed, since `FilePath` is not null here.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs:13]: Isn't it possible to make the `File` property nullable rather than assigning `null!` to this property in the constructor?
I think `file.File.GetText(ctx.CancellationToken)` can be replaced with `file.File?.GetText(ctx.CancellationToken)` in the `SourceGenerator` class.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs]: I'd replace these with `string.Empty`, but not mandatory.

File: README.md
Comment (ycanardeau) [inline README.md]: I'm not a native English speaker, but I guess "by default" would be better?

File: README.md
Comment (ycanardeau) [inline README.md]: here

File: README.md
Comment (ycanardeau) [inline README.md]: and here

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: It's true that I used to use Java many years ago, but I tend to mark classes as sealed as much as possible unless they really need to be inherited (I think classes are sealed by default in newer languages). It's relatively easy to remove the sealed modifier, but hard to add it afterwards (if they have some derived classes already). But it's not a big problem anyway as long as the class is marked as internal.

File: README.md
Comment (ycanardeau) [inline README.md:221]: You need to add these new options to [VocaDb.ResXFileCodeGenerator.props](https://github.com/VocaDB/ResXFileCodeGenerator/blob/main/VocaDb.ResXFileCodeGenerator/build/VocaDb.ResXFileCodeGenerator.props), so that these properties are [visible to the compiler](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#consume-msbuild-properties-and-metadata).

File: README.md
Comment (ycanardeau) [inline README.md:238]: these as well

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: Is it intentional that `Constants.CultureInfoVariable` is appended here?

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs]: This line needs to be negated just like above, otherwise it generates non-static members when I set this option to `true`. It would be nice if you could write unit tests for options classes.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs:12]: Can you please add comments for these options as the number of options has been increasing? Copying and pasting the description from readme should be enough.

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/GlobalOptions.cs:10]: For these too.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: Done, I cleaned it up below, but forgot here

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs:13]: Ofcourse it is. However, it is only ever null in testing, thus my opinion is that it should self document that this cannot ever be null in real code.

File: README.md
Comment (davhdavh) [reply README.md]: They are interchangeable, but sure, fixed

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: done

File: README.md
Comment (davhdavh) [reply README.md:221]: Ahh cool, didn't even know that.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: Nope, good catch, and the unittest was wrong also.

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: Indeed, fixed

File: VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs]: The type name is missing, which causes failing tests. By the way, is it the intended behavior that this property is generated for the "inner" class? Should it be generated for the container class instead as you mentioned in #5? I know you're still editing...

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/FileOptions.cs]: For your information, you can make it internal by using `InternalsVisibleTo`, but I think it's okay to be public for now.

File: VocaDb.ResXFileCodeGenerator/GlobalOptions.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/GlobalOptions.cs]: and this

**Comment by davhdavh** [comment]
It can now generate almost identical code to my 15 year old resx generator plugin, that I had to fix every single visual studio release 👎 
Just need it to support 
* Custom class names (postfix) (Also actually need this one for razor pages)
* Build speed still sucks even after updating to IIncrementalGenerator... It seems it doesn't actually use that to build things faster. I created [https://github.com/dotnet/roslyn/issues/59174](https://github.com/dotnet/roslyn/issues/59174) for this issue.

File: VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator.Tests/GeneratorTests.cs]: Indeed, I made a total mess of that

File: VocaDb.ResXFileCodeGenerator/FileOptions.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/FileOptions.cs]: I know, doesn't matter in this case since its not a library anyone is ever expected to include

**Comment by davhdavh** [comment]
Btw, is there any particular reason why you used `TargetPath` instead of the `Link` that is used by resource generation that comes with msbuild?

**Comment by ycanardeau** [comment]
> Build speed still sucks even after updating to IIncrementalGenerator...

Would it be related to another part of the code rather than roslyn itself? I'm not so confident with my code when it comes to performance...

> Btw, is there any particular reason why you used TargetPath instead of the Link that is used by resource generation that comes with msbuild?

The `TargetPath` option is added by another person in #4, and I don't know the reason why.

**Comment by davhdavh** [comment]
No, I benchmarked it... Just takes ~200msec to generate all 1124 files. It does however take ~13 sec to generate all of the resources files and ~13sec to compile all of the generated cs files and ~3sec to build assembly and copy satellite around for a total of about 30 sec (which is why I moved them to a separate assembly a decade ago, saving those 30 sec every single time I had to rebuild the project has saved me many many hours).

**Comment by ycanardeau** [comment]
@davhdavh This PR is quite big and it will take some time to review it. (I'm glad, and I appreciate you that you've converted this source generator to IIncrementalGenerator.) Do you think it's possible to split it into multiple parts? I think it's best to merge this PR before starting working on version 3.1. By the way, do we need to bump the version number to 3.1 even though we haven't published the nuget package yet?

File: VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj:5]: Is there any particular reason why you changed `LangVersion` to `latest`?

File: VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md:9]: I don't quite understand the purpose of these two files `AnalyzerReleases.Unshipped.md` and `AnalyzerReleases.Shipped.md`. Should I move these rules (adding `## Release 3.1` before `### New Rules`) to `AnalyzerReleases.Shipped.md` *after* publishing a nuget package?

File: README.md
Comment (ycanardeau) [inline README.md:380]: Would you mind if I ask how to benchmark it? (I guess `launchSettings.json` is related to this?) I'm interested in it.

File: VocaDb.ResXFileCodeGenerator.sln.DotSettings
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator.sln.DotSettings:1]: Personally, I haven't used ReSharper, so I'm wondering what the purpose of this file is. Is this different from `.editorconfig`?

**Review by ycanardeau — APPROVED** [review]
Looks good. Is this PR still a work in progress?

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: The `partial` modifier can only appear immediately before `class` ([CS0267](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0267)).

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (ycanardeau) [inline VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: here too

File: VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj:5]: Not really. The latest vs17.5 gave warnings about using preview of .net. And I changed it. Besides using preview versions for something that is used in production isn't a particularly good idea, even if it has no consequences

File: VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md:9]: No. I just used the built-in quick fix which made those files. Don't have a clear idea what they actually do.

File: README.md
Comment (davhdavh) [reply README.md:380]: I used benchmark.net and measured how long it took to do the 2 integration tests

File: VocaDb.ResXFileCodeGenerator.sln.DotSettings
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator.sln.DotSettings:1]: No. Pretty much the same. Although i did autogen the current .editorconfig from resharper, so there shouldn't be any settings that differ. I do highly recommend resharper, it has a ROI measured in days.

File: VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs
Comment (davhdavh) [reply VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs]: Will fix

**Comment by davhdavh** [comment]
I have a small outstanding issue with what to do about dataannotations when not generating static members. But that isn't an issue that should prevent release. And I haven't found a solution yet.

With regards to version numbers, for a project like this it is more about optics than anything else. People might go and read the release notes on a full version than a simple 0.0.1 version bump and see the nice new features.

**Comment by ycanardeau** [comment]
Thanks a lot for the PR!

File: VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md
Comment (DanielBaumert) [reply VocaDb.ResXFileCodeGenerator/AnalyzerReleases.Unshipped.md:9]: "Don't have a clear idea what they actually do." -  Me too.

What I found is https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md but is it helpful? No

**Pull request merged**
